src/Features/CSharp/CodeFixes/Async/CSharpAddAsyncCodeFixProvider.cs

```diff
@@ -62,53 +61,45 @@ protected override string GetDescription(Diagnostic diagnostic, SyntaxNode node,
-            }
-
-            var simpleLambda = oldNode
-                .Ancestors()
-                .FirstOrDefault(n =>
-                    n.IsKind(SyntaxKind.SimpleLambdaExpression));
-
-            if (simpleLambda != null)
+            foreach (var node in oldNode.Ancestors())
             {
-                return simpleLambda;
+                switch (node.Kind())
+                {
+                    case SyntaxKind.ParenthesizedLambdaExpression:
+                    case SyntaxKind.SimpleLambdaExpression:
+                    case SyntaxKind.AnonymousMethodExpression:
+                    case SyntaxKind.MethodDeclaration:
+                        if (node.TypeSwitch(
+                            (ParenthesizedLambdaExpressionSyntax e) => !HasAsyncKeyword(e),
+                            (SimpleLambdaExpressionSyntax e) => !HasAsyncKeyword(e),
+                            (AnonymousMethodExpressionSyntax e) => !HasAsyncKeyword(e),
+                            (MethodDeclarationSyntax e) => !HasAsyncKeyword(e),
+                            (SyntaxNode e) => false))
+                        {
+                            return node;
+                        }
+                        break;
+                    default:
+                        continue;
+                }
             }
 
-            return oldNode
-                .Ancestors()
-                .FirstOrDefault(n =>
-                    n.IsKind(SyntaxKind.MethodDeclaration));
+            return null;
         }
 
-        private async Task<SyntaxNode> ConvertToAsync(SyntaxNode node, SemanticModel semanticModel, Document document, CancellationToken cancellationToken)
-        {
-            var methodNode = node as MethodDeclarationSyntax;
-            if (methodNode != null)
-            {
-                return await ConvertMethodToAsync(document, semanticModel, methodNode, cancellationToken).ConfigureAwait(false);
-            }
-
-            var parenthesizedLambda = node as ParenthesizedLambdaExpressionSyntax;
-            if (parenthesizedLambda != null)
-            {
-                return ConvertParenthesizedLambdaToAsync(parenthesizedLambda);
-            }
-
-            var simpleLambda = node as SimpleLambdaExpressionSyntax;
-            if (simpleLambda != null)
-            {
-                return ConvertSimpleLambdaToAsync(simpleLambda);
-            }
+        private static bool HasAsyncKeyword(ParenthesizedLambdaExpressionSyntax e) => e.AsyncKeyword == SyntaxFactory.Token(SyntaxKind.AsyncKeyword);
+        private static bool HasAsyncKeyword(SimpleLambdaExpressionSyntax e) => e.AsyncKeyword == SyntaxFactory.Token(SyntaxKind.AsyncKeyword);
+        private static bool HasAsyncKeyword(AnonymousMethodExpressionSyntax e) => e.AsyncKeyword == SyntaxFactory.Token(SyntaxKind.AsyncKeyword);
+        private static bool HasAsyncKeyword(MethodDeclarationSyntax e) => e.Modifiers.Any(SyntaxKind.AsyncKeyword);
```

Comment:
Actually, why not just have a single method that takes ```SyntaxNode``` and do the switching here?